text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+  values for each type, which are not nullable)
+* Generated Go interfaces corresponding to a datatype interface would need a suffix, e.g. Iface,  in order to disambiguate it from the struct. This is
+  a bit verbose, and it may be worth considering switching the naming (i.e. adding a suffix to the struct instead) if the interface name is what will
+primarily be used by the customer.
+
+* The alternative to having to implement both a struct and interface is simply translating the datatype interface into a struct. This option has the
+  added advantage of having a more streamlined API, rather than having to convert each property into a getter method and having to call those methods
+to access data fields. For example, something like the
+[HealthCheck](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L596-L642) interface in a container
+definition would be converted into a Go struct:
+
+```go
+type HealthCheck struct {
+    Command []string
+    Interval cdk.Duration;
+    Retries int
+    StartPeriod cdk.Duration;
+    Timeout cdk.Duration;
+}
+```
+
+This would allow a function that takes that interface as an argument, such as
+[renderHealthCheck](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L652-L660) look like this
+(simplified, without accounting for default values):
+
+```go
+func renderHealthCheck(hc HealthCheck) CfnTaskDefinition.HealthCheckProperty {
+    return CfnTaskDefintion.HealthCheckProperty{
+        command:     hc.Command,
+        interval:    hc.Interval,
+        retries:     hc.Retries,
+        startPeriod: hc.StartPeriod,
+        timeout:     hc.Timeout,
+    }
+}
+
+```
+
+Rather than:
+
+```go
+func renderHealthCheck(hc HealthCheck) CfnTaskDefinition.HealthCheckProperty {
+    return CfnTaskDefintion.HealthCheckProperty{
+        command:     hc.GetCommand(),
+        interval:    hc.GetInterval(),
+        retries:     hc.GetRetries(),
+        startPeriod: hc.GetStartPeriod(),
+        timeout:     hc.GetTimeout(),
+    }
+}
+```
+
+## Classes
+
+Typescript takes an object-oriented approach to classes, which includes using polymorphism and subtyping, neither of which are natively supported in
+Go, which is not an object-oriented language. While custom structs, which can be used as pointer receivers in function definitions to simulate
+methods", can be used to encapsulate object-like behavior in Go, subtyping on these custom structs is not possible. In order to simulate subtyping, we
+would need to generate an interface in addition to a concrete struct for each JSII class.
+
+The JSII [ClassType](https://github.com/aws/jsii/blob/master/packages/%40jsii/spec/lib/assembly.ts#L803) provides information on whether a
+class is abstract, whether it extends another class, and whether it implements other interfaces. We will discuss each case in turn.
```

Comment:
I'd like to see an example involving model protected members, and also polymorphic calls. Combining them into one example, how about:


```ts
class Animal {
  public speak() { 
    console.log(this.voice());
  }

  protected voice() {
    return '...';
  }
}

class Poodle extends Animal {
  protected voice() {
    return 'arf';
  }
}

// Expecting to see 'arf'
new Poodle().speak();
```

Reply:
Does the example below showing how `move()` is overridden not fit the bill?

Reply:
Go may need to require "protected" members to be exported as public

Reply:
I think in jsii visibility is already required to be unchanged between inheriting classes (and hopefully releases, TO BE CHECKED), so we could do something like prepend a `_` for protected members.

Reply:
> Does the example below showing how `move()` is overridden not fit the bill?

I don't think so. What I'm looking for is a polymorphic `self` call.

In all other forms of delegation you've shown, a child class is upcalling to its superclass--but that is fine, it knows the type of that superclass.

What if a superclass calls into a virtual method on itself, that may be implemented and overridden by a subclass? You'd need a reference to the current object's interface type, not its struct type.

I mean, I guess there's probably syntax for it, I just don't know what it is.